### ISSUE TITLE
[Bug] `typing_extensions.Protocol` unsupported under `typing_extensions ≥ 4.6.0`

### ISSUE THREAD

**Issue by rsokl** [beartype/beartype#241]
This seems to be specific to `typing_extensions.Protocol` specifically for `typing-extensions==4.6.1`
```python
from beartype import beartype
from typing import Callable, TypeVar
from typing_extensions import Protocol, runtime_checkable

T= TypeVar("T")

@runtime_checkable
class Partial(Protocol[T]):
    __call__: Callable[..., T]


def foo(x: Partial[float]): 
    ...

# raises: beartype.roar.BeartypeDecorHintPepUnsupportedException:
#         function __main__.foo() parameter "x" type hint typing_extensions.Protocol
#         [~T] unsupported but erroneously detected as supported.
beartype(foo)
```

This error is emitted from:

```python
                    # ............{ UNSUPPORTED                        }............
                    # Else, this hint is neither shallowly nor deeply supported and
                    # is thus unsupported. Since an exception should have already
                    # been raised above in this case, this conditional branch
                    # *NEVER* be triggered. Nonetheless, raise an exception.
                    else:
>                       raise BeartypeDecorHintPepUnsupportedException(
                            f'{_EXCEPTION_PREFIX_HINT}'
                            f'{repr(hint_curr)} unsupported but '
                            f'erroneously detected as supported.'
                        )
```
so I figured I should report this :)

**Comment by leycec** [comment]
Thanks so much for all the fine deets on how @beartype has let us all down. :sweat_smile: 

More importantly, though... it's *really* fantastic to hear from you again, @rsokl! I love surreptitiously following the upward trajectory of my favourite users – and your trajectory is ever-upward. You've accumulated a career's worth of utterly fascinating Python-friendly projects in the scant time I've known you. Your legend is legendary.

## @beartype: Back in Black (Fur)

**Okay.** Back to our shameful failures. `typing_extensions` is non-trivial (albeit critical) to support at runtime, because the type hints it exposes tend to differ from the runtime APIs of equivalent type hints in the standard `typing` module. But...

It kinda doesn't matter who's to blame in the end, right? It's still ultimately @beartype's fault. We support `typing_extensions`. So, we just gotta do this.

Thankfully, I have an ~~awful~~ useful interim solution until we do: **please use `typing.Protocol` rather than `typing_extensions.Protocol`.** I know; I know. Super-lame. But Python 3.7 is on the cusp of hitting its official End-Of-Life (EOL). Everything (including @beartype) will be dropping support for Python 3.7 shortly. Since `typing.Protocol` was first introduced by Python 3.8.0, **`typing.Protocol` should now be safely importable everywhere.**

There's basically no demonstrable benefit to importing `typing_extensions.Protocol` anymore. Of course, @beartype should still resolve this directly for the good of Python- and LLaMa-kind.

**Comment by AlexWaygood** [comment]
FYI, all the changes that have been made to `typing_extensions.Protocol` in v4.6.0 are also going to be included in `typing.Protocol` in Python 3.12. The changes (to both `typing.Protocol` and `typing_extensions.Protocol`) were mostly made in order to achieve a 20x speedup for `isinstance()` checks against runtime-checkable protocols.

`typing_extensions` also backports several significant bugfixes that have been made to `typing.Protocol` in recent Python versions, but haven't been backported as far as Python 3.8.

**Comment by AlexWaygood** [comment]
Here's a rundown of significant changes that have been made to `typing.Protocol` in Python 3.12. Without knowing which `typing` implementation details you're looking at in beartype, not sure which of these are likely to affect you or not:

- https://github.com/python/cpython/pull/103034
- https://github.com/python/cpython/pull/103141
- https://github.com/python/cpython/pull/103152
- https://github.com/python/cpython/pull/103159
- https://github.com/python/cpython/pull/103160
- https://github.com/python/cpython/pull/103280
- https://github.com/python/cpython/pull/104622
- https://github.com/python/cpython/pull/104556
- https://github.com/python/cpython/pull/104559

There's also documentation on the upcoming changes in 3.12 here: https://docs.python.org/3.12/whatsnew/3.12.html#typing.

**Comment by leycec** [comment]
**WOW.** Thanks so much for that full-send dissection of `typing{_extension,}.Protocol` micro- and macro-optimizations, @AlexWaygood. Everything you listed is wonderful to see; this will give user-defined protocols a big boost with respect to runtime efficiency, community adoption, and feature parity with static type-checkers. I can only parrot @rsokl:

> That rocks!!

Thankfully, this specific issue is almost certainly a trivial misstep on @beartype's end. Our test suite failed to exhaustively test all `typing_extensions` attributes. Now, we all pay the QA price. :face_exhaling:

**Comment by leycec** [comment]
**Rejoice!** I've escalated this to a priority issue in my head. Hilariously, I appear to have known about this several years ago and then promptly ignored it by explicitly excluding `typing_extensions.Protocol` from testing. <sup>*...actually, that's not funny.*</sup>

In our voluminous test suite that exhausts my soul whenever I peek inside it, I just discovered this ominous phrasing:

```
# Note this iteration intentionally excludes the non-standard third-party
# "typing_extensions.Protocol" superclass, whose runtime behaviour has been
# shown to diverge significantly from that of standard implementations.
```

A pox on you, younger @leycec! May cats meow in the night unconsolably until I *finally* resolve this. :crying_cat_face:

**Comment by leycec** [comment]
**Curioser** and **curioser.** I've trivially re-enabled testing support for `typing_extensions.Protocol`, for which we can all be thankful.

The good news is that these tests now pass under `typing_extensions < 4.6.0`! So, @beartype now officially supports older implementations of `typing_extensions.Protocol`.

The bad news is that I don't know why these tests purportedly fail under `typing_extensions ≥ 4.6.0`. I smell a code smell... *and it smells.* :wilted_flower:

**Comment by leycec** [comment]
**Ah-ha!** After a week of relentless video gaming that has worn my thumbs to stubs, I *finally* investigated this tonight and can explain what is happening here. Pump those fists! Under:

* Prior versions of `typing` and `typing_extensions`, @beartype detected **protocols** (i.e., `typing{_extensions,}.Protocol` subclasses) as **generics** (i.e., `typing{_extensions,}.Generic` subclasses). @beartype handled both using the exact same logic, because they were literally the exact same thing. Protocols and generics only *appeared* to be superficially different things from the end user perspective.
* Newer versions of `typing` and `typing_extensions`, @beartype now detects protocols and generics as distinct kinds of things. This is a good thing. Although protocols *are* generics, they're a bit more than *just* generics. Right?

That said, the entirety of the @beartype codebase still crudely assumes that protocols are literally just generics in disguise. I'll need to tinker a bit in the steaming innards of @beartype to unwind this bad assumption. Let's see if I do something other than relentless video gaming for once. :sweat_smile:

**Comment by leycec** [comment]
**Resolved by 8e318a8.** Thanks for your effusive patience, wonderful userbase who are about to be rocked by the newly optimized `typing_extensions.Protocol` class. Likewise, thanks to @AlexWaygood for the deep dive into exactly what these optimizations are and why we all need them in our lives *now.*

Since `typing_extensions` support is critical, I'll be releasing this shortly with our upcoming `beartype 0.14.2` patch release. Praise be to the Pythonistas. We drive humanity ever forward.
